usr.bin/nl/nl.1
.\"	$NetBSD$
.\"
.\" Copyright (c) 1999 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Klaus Klein.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd May 1, 2013
.Dt NL 1
.Os
.Sh NAME
.Nm nl
.Nd line numbering filter
.Sh SYNOPSIS
.Nm
.Op Fl p
.Op Fl b Ar type
.Op Fl d Ar delim
.Op Fl f Ar type
.Op Fl h Ar type
.Op Fl i Ar incr
.Op Fl l Ar num
.Op Fl n Ar format
.Op Fl s Ar sep
.Op Fl v Ar startnum
.Op Fl w Ar width
.Op Ar file
.Sh DESCRIPTION
The
.Nm
utility reads lines from the named
.Ar file ,
applies a configurable line numbering filter operation,
and writes the result to the standard output.
If
.Ar file
is a single dash
.Pq Sq \&-
or absent,
.Nm
reads from the standard input.
.Pp
The
.Nm
utility treats the text it reads in terms of logical pages.
Unless specified otherwise, line numbering is reset at the start of each
logical page.
A logical page consists of a header, a body and a footer section; empty
sections are valid.
Different line numbering options are independently available for header,
body and footer sections.
.Pp
The starts of logical page sections are signaled by input lines containing
nothing but one of the following sequences of delimiter characters:
.Bl -column "\e:\e:\e: " "header " -offset indent
.It Em "Line" Ta Em "Start of"
.It \e:\e:\e:	header
.It \e:\e:	body
.It \e:	footer
.El
.Pp
If the input does not contain any logical page section signaling directives,
the text being read is assumed to consist of a single logical page body.
.Pp
The following options are available:
.Bl -tag -width indent
.It Fl b Ar type
Specify the logical page body lines to be numbered.
Recognized
.Ar type
arguments are:
.Bl -tag -width pstringXX
.It a
Number all lines.
.It t
Number only non-empty lines.
.It n
No line numbering.
.It p Ns Ar expr
Number only those lines that contain the basic regular expression specified
by
.Ar expr .
.El
.Pp
The default
.Ar type
for logical page body lines is t.
.It Fl d Ar delim
Specify the delimiter characters used to indicate the start of a logical
page section in the input file.
At most two characters may be specified; if only one character is specified,
the first character is replaced and the second character remains unchanged.
The default
.Ar delim
characters are ``\e:''.
.It Fl f Ar type
Specify the same as
.Fl b Ar type
except for logical page footer lines.
The default
.Ar type
for logical page footer lines is n.
.It Fl h Ar type
Specify the same as
.Fl b Ar type
except for logical page header lines.
The default
.Ar type
for logical page header lines is n.
.It Fl i Ar incr
Specify the increment value used to number logical page lines.
The default
.Ar incr
value is 1.
.It Fl l Ar num
If numbering of all lines is specified for the current logical section
using the corresponding
.Fl b
a,
.Fl f
a
or
.Fl h
a
option,
specify the number of adjacent blank lines to be considered as one.
For example,
.Fl l
2 results in only the second adjacent blank line being numbered.
The default
.Ar num
value is 1.
.It Fl n Ar format
Specify the line numbering output format.
Recognized
.Ar format
arguments are:
.Pp
.Bl -tag -width lnXX -compact -offset indent
.It ln
Left justified.
.It rn
Right justified, leading zeros suppressed.
.It rz
Right justified, leading zeros kept.
.El
.Pp
The default
.Ar format
is rn.
.It Fl p
Specify that line numbering should not be restarted at logical page delimiters.
.It Fl s Ar sep
Specify the characters used in separating the line number and the corresponding
text line.
The default
.Ar sep
setting is a single tab character.
.It Fl v Ar startnum
Specify the initial value used to number logical page lines; see also the
description of the
.Fl p
option.
The default
.Ar startnum
value is 1.
.It Fl w Ar width
Specify the number of characters to be occupied by the line number;
if the
.Ar width
is insufficient to hold the line number, it will be truncated to its
.Ar width
least significant digits.
The default
.Ar width
is 6.
.El
.Sh EXIT STATUS
.Ex -std
.Sh SEE ALSO
.Xr pr 1
.Sh STANDARDS
The
.Nm
utility is compliant with the
.St -xpg4.2
specification.
.Sh HISTORY
The
.Nm
utility first appeared in
.At V.2 .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 6.0 to NetBSD 7.0:
	window(1): Removed from the base system in favor of tmux(1).  Use the
		pkgsrc/misc/window package from now on.  [jmmv 20120215]
	pkg_add(1): imported 20120221 version from pkgsrc [wiz 20120221]
	file(1): upgraded to 5.11 [christos 20120222]
	evbmips: Add support for Lemote Yeeloong Notebook. [nonaka 20120302]
	dbsym(8): Add option -p [bsh 20120319]
	dhcpcd(8): Import dhcpcd-5.5.5 [roy 20120323]
	ipf: upgrade to 5.1.1 with new build framework, no module and
		test support yet [christos 20120323]
	pcc: upgrade to 1.1.0.DEVEL 20120325 [plunky 20120325]
	openresolv(8): Import openresolv-3.5.1 [roy 20120327]
	dhcpcd(8): Import dhcpcd-5.5.6 [roy 20120328]
	apple: Recognize the apple disk partitioning scheme, and auto-discover
		wedges. Fix iPod disk mounting, by adding quirk.
		[christos 20120306]
	fstab: Add NAME=wedgename for fs_spec and routines to parse it. Now
		wedges can be used in /etc/fstab, even in dynamic environments
		where disks come and go and wedges get renumbered.
		[christos 20120306]
	pdisk(8): Make it work for 64 bit hosts, and make it MI.
		[christos 20120306]
	amiga: Add clockport(4) layer and a driver for A1200 on-board clockport
		a1k2cp(4). [rkujawa 20120417]
	kernel: add mpii(4), a driver for LSI Logic Fusion-MPT Message Passing
		Interface II SAS controllers.  [bouyer 20120419]
	kernel: Add support for 256 CPUs on amd64. [rmind 20120420]
	OpenSSH: Imported 6.0 [christos 20120501]
	bzip2: Imported 1.0.6 [wiz 20120507]
	amiga: Rework X-Surf driver as xsurf(4). Add support for clockport(4)
		on top of xsurf(4). [rkujawa 20120515]
	backtrace(3): Add <execinfo.h> and the backtrace family of functions
		[christos 20120526]
	kernel: Add run(4), a driver for Ralink Technology
		RT2700U/RT2800U/RT3000U USB IEEE 802.11a/b/g/n wireless
		network devices, ported from OpenBSD. [nonaka 20120530]
	bind: update to 9.9.1-P1 to fix the recent rdata exploit.
		[christos 20120604]
	empb(4): Add ELBOX Mediator PCI 1200 driver. [rkujawa 20120604]
	kernel: Add a new RLIMIT_NTHR limit to limit the number of processes
		per user and the corresponding kern.maxlwp sysctl. Expose
		kern.uidinfo.* via sysctl. [christos 20120608]
	usb(9): Added support for SMP to the USB subsytem.  By Jared D.
		McNeill and myself.  [mrg 20120609]
	kernel: {send,recv}mmsg support [christos 20120621]
	kernel: Add sysctls to avoid ipv6 DoS attacks (from OpenBSD):
		net.inet6.ip6.neighborgcthresh = 2048
		net.inet6.ip6.maxifprefixes = 16
		net.inet6.ip6.maxifdefrouters = 16
		net.inet6.ip6.maxdynroutes = 4096
		[christos 20120622]
	ata(4): added support for sata Port MultiPliers (PMP)
		[bouyer 20120702]
	kernel, libc: Add MurmurHash2 function. [rmind 20120708]
	atf(7): Import 0.16.  [jmmv 20120711]
	dhcpcd(8): Import dhcpcd-5.6.1 [roy 20120712]
	kernel: Add support for sensors to provide entropy to rnd(4)
		[pgoyette 20120715]
	tdvfb(4): Add 3Dfx Voodoo2 driver. [rkujawa 20120719]
	tdvfb(4): Add 3Dfx Voodoo Graphics (aka Voodoo1) support.
		[rkujawa 20120720]
	kernel: Change mii_statchg to take "struct ifnet *" instead of
		"device_t". [matt 20120722]
	pr(1): Add support for POSIX -f and -p options. [ginsbach 20120724]
	powerpc: Add support for FPU emulation on BookE. [matt 20120722]
	OpenSSL: Imported 1.0.1c [christos 20120726]
	kernel: safepri is dead; replaced by macro IPL_SAFEPRI. [matt 20120727]
	kernel: Add malo(4), a driver for Marvell Libertas IEEE 802.11b/g
		wireless network devices, ported from OpenBSD.
		[degroote 20120730]
	kernel, libc: Add modctl(MODCTL_EXISTS, ...) to determine if modules
		can be loaded. [jnemeth 20120803]
	modstat(8): Add -A, -a, and -e options for testing module loadability.
		[jnemeth 20120803]
	libc: Update to tzcode2012e. [christos 20120809]
	arm: Add support for BEAGLEBOARDXM (TI DM37XX cpus) boards.
		[matt 20120822]
	mfi(8): Added tagged queuing support.  [bouyer 20120823]
	mfi(8): Added support for LSI SAS2208-based controllers.
		[bouyer 20120823]
	mkdep(1): Add -P option.  [jmmv 20120826]
	share/mk: Add placeholder test programs (with some very basic test
		cases within) for common bsd.*.mk modules.  [jmmv 20120826]
	arm: Add Cortex-A9 support (a9tmr, a9wdt, armgic, armperiph).  Add
		initial MP support (can boot 2nd cpu but dies quickly).  Add
		L1 SuperSection support for ARMv6/ARMv7.  Add support for 
		BCM5301X SoC.  Add initial support for OMAP4430 and partial
		support for Freescale i.MX6.  Add new common boot/kvminit
		code.  [matt 20120831]
	dhcpcd(8): Import dhcpcd-5.6.2 [roy 20120903]
	bge(4): Add support for BCM57762, found in Apple's Thunderbolt to
		Gigabit Ethernet Adapter. PR/46961 [tsutsui 20120917]
	mfi(4): add a command pass-through ioctl, and associated COMPAT_LINUX
		support. This allows to use the LSI MegaCLI linux binary
		on a NetBSD host.  [bouyer 20120919]
	zoneinfo: Import tzdata2012f. [apb 20120927]
	kernel: add clock_nanosleep(2) [christos 20121001]
	wpa:	New wpa_supplicant(8) and hostapd(8) 1.0 [christos 20121007]

usr.bin/nl/nl.c
/*	$NetBSD$	*/

/*-
 * Copyright (c) 1999 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Klaus Klein.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1999\
 The NetBSD Foundation, Inc.  All rights reserved.");
__RCSID("$NetBSD$");
#endif    

#include <errno.h>
#include <limits.h>
#include <locale.h>
#include <regex.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <err.h>

typedef enum {
	number_all,		/* number all lines */
	number_nonempty,	/* number non-empty lines */
	number_none,		/* no line numbering */
	number_regex		/* number lines matching regular expression */
} numbering_type;

struct numbering_property {
	const char * const	name;		/* for diagnostics */
	numbering_type		type;		/* numbering type */
	regex_t			expr;		/* for type == number_regex */
};

/* line numbering formats */
#define FORMAT_LN	"%-*d"	/* left justified, leading zeros suppressed */
#define FORMAT_RN	"%*d"	/* right justified, leading zeros suppressed */
#define FORMAT_RZ	"%0*d"	/* right justified, leading zeros kept */

#define FOOTER		0
#define BODY		1
#define HEADER		2
#define NP_LAST		HEADER

static struct numbering_property numbering_properties[NP_LAST + 1] = {
	{ "footer",	number_none,	{ 0, 0, 0, 0 } },
	{ "body",	number_nonempty, { 0, 0, 0, 0 } },
	{ "header",	number_none,	{ 0, 0, 0, 0 } },
};

#define max(a, b)	((a) > (b) ? (a) : (b))

/*
 * Maximum number of characters required for a decimal representation of a
 * (signed) int; courtesy of tzcode.
 */
#define INT_STRLEN_MAXIMUM \
	((sizeof (int) * CHAR_BIT - 1) * 302 / 1000 + 2)

static void	filter(void);
static void	parse_numbering(const char *, int);
static void	usage(void) __attribute__((__noreturn__));

/*
 * Pointer to dynamically allocated input line buffer, and its size.
 */
static char *buffer;
static size_t buffersize;

/*
 * Dynamically allocated buffer suitable for string representation of ints.
 */
static char *intbuffer;
static size_t intbuffersize;

/*
 * Configurable parameters.
 */
/* delimiter characters that indicate the start of a logical page section */
static char delim[2] = { '\\', ':' };

/* line numbering format */
static const char *format = FORMAT_RN;

/* increment value used to number logical page lines */
static int incr = 1;

/* number of adjacent blank lines to be considered (and numbered) as one */
static unsigned int nblank = 1;

/* whether to restart numbering at logical page delimiters */
static int restart = 1;

/* characters used in separating the line number and the corrsp. text line */
static const char *sep = "\t";

/* initial value used to number logical page lines */
static int startnum = 1;

/* number of characters to be used for the line number */
/* should be unsigned but required signed by `*' precision conversion */
static int width = 6;


int
main(int argc, char *argv[])
{
	int c;
	long val;
	unsigned long uval;
	char *ep;

	(void)setlocale(LC_ALL, "");

	/*
	 * Note: this implementation strictly conforms to the XBD Utility
	 * Syntax Guidelines and does not permit the optional `file' operand
	 * to be intermingled with the options, which is defined in the
	 * XCU specification (Issue 5) but declared an obsolescent feature that
	 * will be removed from a future issue.  It shouldn't matter, though.
	 */
	while ((c = getopt(argc, argv, "pb:d:f:h:i:l:n:s:v:w:")) != -1) {
		switch (c) {
		case 'p':
			restart = 0;
			break;
		case 'b':
			parse_numbering(optarg, BODY);
			break;
		case 'd':
			if (optarg[0] != '\0')
				delim[0] = optarg[0];
			if (optarg[1] != '\0')
				delim[1] = optarg[1];
			/* at most two delimiter characters */
			if (optarg[2] != '\0') {
				errx(EXIT_FAILURE,
				    "invalid delim argument -- %s",
				    optarg);
				/* NOTREACHED */
			}
			break;
		case 'f':
			parse_numbering(optarg, FOOTER);
			break;
		case 'h':
			parse_numbering(optarg, HEADER);
			break;
		case 'i':
			errno = 0;
			val = strtol(optarg, &ep, 10);
			if ((ep != NULL && *ep != '\0') ||
			 ((val == LONG_MIN || val == LONG_MAX) && errno != 0))
				errx(EXIT_FAILURE,
				    "invalid incr argument -- %s", optarg);
			incr = (int)val;
			break;
		case 'l':
			errno = 0;
			uval = strtoul(optarg, &ep, 10);
			if ((ep != NULL && *ep != '\0') ||
			    (uval == ULONG_MAX && errno != 0))
				errx(EXIT_FAILURE,
				    "invalid num argument -- %s", optarg);
			nblank = (unsigned int)uval;
			break;
		case 'n':
			if (strcmp(optarg, "ln") == 0) {
				format = FORMAT_LN;
			} else if (strcmp(optarg, "rn") == 0) {
				format = FORMAT_RN;
			} else if (strcmp(optarg, "rz") == 0) {
				format = FORMAT_RZ;
			} else
				errx(EXIT_FAILURE,
				    "illegal format -- %s", optarg);
			break;
		case 's':
			sep = optarg;
			break;
		case 'v':
			errno = 0;
			val = strtol(optarg, &ep, 10);
			if ((ep != NULL && *ep != '\0') ||
			 ((val == LONG_MIN || val == LONG_MAX) && errno != 0))
				errx(EXIT_FAILURE,
				    "invalid startnum value -- %s", optarg);
			startnum = (int)val;
			break;
		case 'w':
			errno = 0;
			val = strtol(optarg, &ep, 10);
			if ((ep != NULL && *ep != '\0') ||
			 ((val == LONG_MIN || val == LONG_MAX) && errno != 0))
				errx(EXIT_FAILURE,
				    "invalid width value -- %s", optarg);
			width = (int)val;
			if (!(width > 0))
				errx(EXIT_FAILURE,
				    "width argument must be > 0 -- %d",
				    width);
			break;
		case '?':
		default:
			usage();
			/* NOTREACHED */
		}
	}
	argc -= optind;
	argv += optind;

	switch (argc) {
	case 0:
		break;
	case 1:
		if (strcmp(argv[0], "-") != 0 &&
		    freopen(argv[0], "r", stdin) == NULL)
			err(EXIT_FAILURE, "Cannot open `%s'", argv[0]);
		break;
	default:
		usage();
		/* NOTREACHED */
	}

	/* Determine the maximum input line length to operate on. */
	if ((val = sysconf(_SC_LINE_MAX)) == -1) /* ignore errno */
		val = LINE_MAX;
	/* Allocate sufficient buffer space (including the terminating NUL). */
	buffersize = (size_t)val + 1;
	if ((buffer = malloc(buffersize)) == NULL)
		err(EXIT_FAILURE, "Cannot allocate input line buffer");

	/* Allocate a buffer suitable for preformatting line number. */
	intbuffersize = max((int)INT_STRLEN_MAXIMUM, width) + 1; /* NUL */
	if ((intbuffer = malloc(intbuffersize)) == NULL)
		err(EXIT_FAILURE, "cannot allocate preformatting buffer");

	/* Do the work. */
	filter();

	return EXIT_SUCCESS;
	/* NOTREACHED */
}

static void
filter(void)
{
	int line;		/* logical line number */
	int section;		/* logical page section */
	unsigned int adjblank;	/* adjacent blank lines */
	int consumed;		/* intbuffer measurement */
	int donumber, idx;

	adjblank = 0;
	line = startnum;
	section = BODY;
#ifdef __GNUC__
	donumber = 0;	/* avoid bogus `uninitialized' warning */
#endif

	while (fgets(buffer, (int)buffersize, stdin) != NULL) {
		for (idx = FOOTER; idx <= NP_LAST; idx++) {
			/* Does it look like a delimiter? */
			if (buffer[2 * idx + 0] == delim[0] &&
			    buffer[2 * idx + 1] == delim[1]) {
				/* Was this the whole line? */
				if (buffer[2 * idx + 2] == '\n') {
					section = idx;
					adjblank = 0;
					if (restart)
						line = startnum;
					goto nextline;
				}
			} else {
				break;
			}
		}

		switch (numbering_properties[section].type) {
		case number_all:
			/*
			 * Doing this for number_all only is disputable, but
			 * the standard expresses an explicit dependency on
			 * `-b a' etc.
			 */
			if (buffer[0] == '\n' && ++adjblank < nblank)
				donumber = 0;
			else
				donumber = 1, adjblank = 0;
			break;
		case number_nonempty:
			donumber = (buffer[0] != '\n');
			break;
		case number_none:
			donumber = 0;
			break;
		case number_regex:
			donumber =
			    (regexec(&numbering_properties[section].expr,
			    buffer, 0, NULL, 0) == 0);
			break;
		}

		if (donumber) {
			consumed = snprintf(intbuffer, intbuffersize, format,
			    width, line);
			(void)printf("%s",
			    intbuffer + max(0, consumed - width));
			line += incr;
		} else {
			(void)printf("%*s", width, "");
		}
		(void)printf("%s%s", sep, buffer);

		if (ferror(stdout))
			err(EXIT_FAILURE, "output error");
nextline:
		;
	}

	if (ferror(stdin))
		err(EXIT_FAILURE, "input error");
}

/*
 * Various support functions.
 */

static void
parse_numbering(const char *argstr, int section)
{
	int error;
	char errorbuf[NL_TEXTMAX];

	switch (argstr[0]) {
	case 'a':
		numbering_properties[section].type = number_all;
		break;
	case 'n':
		numbering_properties[section].type = number_none;
		break;
	case 't':
		numbering_properties[section].type = number_nonempty;
		break;
	case 'p':
		/* If there was a previous expression, throw it away. */
		if (numbering_properties[section].type == number_regex)
			regfree(&numbering_properties[section].expr);
		else
			numbering_properties[section].type = number_regex;

		/* Compile/validate the supplied regular expression. */
		if ((error = regcomp(&numbering_properties[section].expr,
		    &argstr[1], REG_NEWLINE|REG_NOSUB)) != 0) {
			(void)regerror(error,
			    &numbering_properties[section].expr,
			    errorbuf, sizeof (errorbuf));
			errx(EXIT_FAILURE,
			    "%s expr: %s -- %s",
			    numbering_properties[section].name, errorbuf,
			    &argstr[1]);
		}
		break;
	default:
		errx(EXIT_FAILURE,
		    "illegal %s line numbering type -- %s",
		    numbering_properties[section].name, argstr);
		exit(EXIT_FAILURE);
	}
}

static void
usage(void)
{
	(void)fprintf(stderr, "Usage: %s [-p] [-b type] [-d delim] [-f type] "
	    "[-h type] [-i incr] [-l num]\n\t[-n format] [-s sep] "
	    "[-v startnum] [-w width] [file]\n", getprogname());
	exit(EXIT_FAILURE);
}
